libpromises/eval_context.c

```diff
@@ -3847,3 +3902,180 @@ bool EvalContextOverrideImmutableGet(EvalContext *ctx)
     assert(stack_frame != NULL);
     return stack_frame->override_immutable;
 }
+
+// ##############################################################
+
+static EventFrame *EventFrameCreate(char *type, const char *name, const char *source, SourceOffset offset)
+{
+    EventFrame *event = (EventFrame *) xmalloc(sizeof(EventFrame));
+
+    event->type = type;
+    event->name = SafeStringDuplicate(name);
+    char *source_path = GetAbsolutePath(source);
+    event->filename = (source_path == NULL) ? SafeStringDuplicate("") : source_path;
+    event->elapsed = 0;
+    event->children = SeqNew(10, NULL);
+    event->offset = offset;
+    event->parent = NULL;
+    event->id = StringFormat("%s_%s_%s_%ld_%ld", type, name, source, offset.start, offset.line);
+    
+    return event;
+}
+
+static void EventFrameDestroy(EventFrame *event)
+{
+    assert(event != NULL);
+    SeqDestroy(event->children);
+    free(event->filename);
+    free(event->id);
+    free(event->name);
+    free(event);
+}
+
+static int EventFrameCompare(const void *event1, const void *event2, ARG_UNUSED void *user_data)
+{
+    assert(event1 != NULL);
+    assert(event2 != NULL);
+    const EventFrame *e1 = event1;
+    const EventFrame *e2 = event2;
+    return strcmp(e1->id, e2->id);
+}
+
+static void EventFrameUpdate(EvalContext *ctx, EventFrame *last_event)
+{
+    assert(ctx != NULL);
+    assert(last_event != NULL);
+    SeqAppend(ctx->events, last_event); // keep track of EventFrames for cleanup
```

Comment:
@victormlg I think this part (pushing an event into the Seq) is the only thing we want to do while evaluating policy. The rest (building a nice tree, hashmaps, summarizing, sorting, etc. Should all appen by iterating over the events after policy evaluation is done.